src/matcha_ml/templates/azure_template/azure_template.py

```diff
@@ -0,0 +1,98 @@
+"""Build a template for provisioning resources on Azure using terraform files."""
+import json
+import os
+from typing import Optional
+
+import typer
+
+from matcha_ml.cli._validation import check_current_deployment_exists
+from matcha_ml.cli.ui.print_messages import print_status
+from matcha_ml.state import MatchaStateService
+from matcha_ml.templates.base_template import (
+    BaseTemplate,
+    TemplateVariables,
+)
+
+SUBMODULE_NAMES = [
+    "aks",
+    "resource_group",
+    "mlflow_module",
+    "storage",
+    "seldon",
+    "zenml_storage",
+    "zen_server",
+    "azure_container_registry",
+    "zen_server/zenml_helm",
+    "zen_server/zenml_helm/templates",
+]
+
+
+class AzureTemplate(BaseTemplate):
+    """A template tailored for provisioning the resources on azure.
+
+    Inherits:
+        BaseTemplate: The base template class.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the StateStorageTemplate with the submodule names.
+
+        Args:
+            submodule_names (List[str]): A list of submodule names.
+        """
+        super().__init__(SUBMODULE_NAMES)
+
+    def reuse_configuration(self, path: str) -> bool:
+        """Check if a configuration already exists, and prompt user to override or reuse it.
+
+        Args:
+            path (str): path to the infrastructure configuration
+
+        Returns:
+            bool: decision to reuse the existing configuration
+        """
+        if os.path.exists(path):
+            if check_current_deployment_exists():
+                matcha_state_service = MatchaStateService()
+                resource_group_name = (
+                    matcha_state_service.fetch_resources_from_state_file(
+                        "cloud", "prefix"
+                    )["cloud"]["prefix"]
+                )
+                warning_msg = f"\nWARNING: Matcha has detected that a deployment already exists in Azure with the resource group name '{resource_group_name}'. Use 'matcha destroy' to remove these resources before trying to provision."
+                confirmation_msg = "\nIf you continue, you will create a orphan resource. You should destroy the resources before proceeding.\n\nDo you want to override the existing configuration?"
```

Comment:
We should probably update this, 

In the case that someone runs `matcha destroy` and then `matcha provision` with the same state/resource group it means that they will encounter this message when they just want to provision the other resources.

Does it still make sense to have this?

Reply:
This may actually apply to the whole `def reuse_configuration(self, path: str) -> bool:` function

Reply:
If someone runs `matcha destroy`, then `check_current_deployment_exists()` would return `False`, then it should return the message after `else`?

Reply:
If someone runs `matcha destroy` they will still have the resource group and therefore `check_current_deployment_exists` will return True.

Reply:
I thought `check_current_deployment_exists` should check whether the resources in `matcha.state` are provisioned on azure.

Reply:
It checks if a resource group exists for the resource group name found in the `matcha.state` file

Reply:
I don't think it should check whether the resource group exists in `check_current_deployment_exists` as resource group are created by the remote state storage infrastructure and `matcha provision` would already checked if a remote state storage exists before running the `azure_template`.

@d-lowl Do [RPD-163](https://fuzzy-labs.atlassian.net/jira/software/projects/RPD/boards/4?selectedIssue=RPD-163&text=163) check whether a resource group exists or just assume it exists based on the `matcha.config.json`.

Reply:
Might be wrong, but we should check if resource group exists in `is_state_provisioned` and `reuse_configuration` should check whether the resources in `infrastructure/resources` are provisioned. 

I can think of a few scenarios:
- If resources are not provisioned, and `.matcha/infrastructure/resources` does not exist, provision the resources.
- If resources are not provisioned but `.matcha/infrastructure/resources` exist, prompt user the message `Matcha has detected that the you already have resources configured for provisioning` and ask if they would like to overide.
- If resources are provisioned, then do not re-provision and tell the user that resources are already provision by another user.

cc'ing @dudeperf3ct @d-lowl for some inputs.

Reply:
Indeed, currently matcha does not check if resources are actually provisioned (we just run terraform apply, without looking at the plan). The warning message is indeed confusing: even if we are running apply over existing resources with the same names (hence same tf configuration), it will succeed, there's no need to run destroy. It will fail if someone else has provisioned a resource group with the same name by other means though.

Reply:
>If resources are provisioned, then do not re-provision and tell the user that resources are already provision by another user.

We also need to keep in mind that terraform configuration might be only partially applied (e.g. terraform apply has been run, but something was added afterwards). What we do want to detect in fact if the prefix was used/resource group has been provisioned, but by something else, rather then this matcha instance. It's trickier to implement, and should probably be a ticket of its own

Reply:
The remote state manager does check whether the resource group exists implicitly, when the connection string is created I think @dudeperf3ct ? (it does not check if it was provisioned by matcha)

Reply:
>The remote state manager does check whether the resource group exists implicitly, when the connection string is created I think

Yes, the check is implicit that is we use try, except block to raise a error. I think we need a more explicit check on whether resource group exists and then we should use that resource group. Ideally,  I would include it in `is_state_provisioned` [function](https://github.com/fuzzylabs/matcha/blob/develop/src/matcha_ml/state/remote_state_manager.py#L132), before we check if storage container exists.

https://github.com/fuzzylabs/matcha/blob/c7b0a6c373c5b5c9b31fb5cf24cf238c573ef39b/src/matcha_ml/services/azure_service.py#L197-L221